routes/users.js

```diff
@@ -0,0 +1,114 @@
+const express = require("express");
+const db = require("../database");
+
+const router = express.Router();
+
+router.get("/:user_id", (req, res) => {
+  db("users")
+    .where("id", req.params.user_id)
+    .then(results => {
+      if (results.length === 0) {
+        res.status(500).json({ message: "User not found" });
+      } else {
+        res.json(results[0]);
+      }
+    })
+    .catch(err => {
+      res.status(500).json({ message: err.message });
+    });
+});
+
+router.get("/", (req, res) => {
+  db("users")
+    .then(results => {
+      res.json(results);
+    })
+    .catch(err => {
+      res.status(500).json({ message: err.message });
+    });
+});
+
+router.post("/login", (req, res) => {
+  db("users")
+    .where("email", req.body.email)
+    .then(results => {
+      if (results.length === 0) {
+        res.status(500).json({ message: "User not found" });
+      } else if (results[0].password !== req.body.password) {
+        res.status(500).json({ message: "Incorrect password" });
+      } else res.json(results[0]);
+    })
+    .catch(err => {
+      res.status(500).json({ message: err.message });
+    });
+});
+
+router.post("/register", (req, res) => {
+  db("users")
+    .where("email", req.body.email)
+    .then(results => {
+      if (results.length > 0) {
+        res.status(500).json({ message: "User already exists" });
+      } else if (!req.body.email || !req.body.password) {
+        res.status(500).json({ message: "Must POST email and password" });
+      } else {
+        db("users")
```

Comment:
Do not nest promises, return them instead so that they are part of the parent promise.

Reply:
Use the errors to your advantage here. Don't see if a user exists first just try to insert one and work with any errors you receive.

Reply:
The database accepts empty string values from the request even though the email and password columns have notNullable constraints, so how can I get it to throw an error for empty string values without coding for it? And when I do get an error, like when the email column's unique constraint is triggered and produces the error "insert into \"users\" (\"email\", \"password\") values ($1, $2) returning * - duplicate key value violates unique constraint \"users_email_unique\"", how do I identify and replace it with the required JSON object { "message": "User already exists" } ?

Reply:
So the easiest way to handle this is don't even go to the database if either of those are empty strings.

Reply:
I've pushed a new commit that removes the nesting, except for transactions, apparently some nesting is part of the transaction pattern? Can you review the new routes? Thanks.